### DISCUSSION TITLE
FOUC problem in RTL

### DISCUSSION THREAD

**Discussion by ehsanonline** [mantinedev/mantine#1118] [Q&A] [answered]
I'm using Mantine with Nextjs and in production build, while the page is loading, elements don't have styles until some noticeable time of loading, even with good connection. After some research I found that this problem called FOUC (Flash Of Unstyled Component). For example the `<Button/>` while loading is just a regular browser button with no style, the HTML loaded but styles don't; Or the `<Center>` contents are not at the center while page is loading **which makes bad UX.** All other attributes like `px` and... have the same problem. but `style` ,`style-jsx` , global and module styles are fine.
My project is very small (just one light page). I don't want to use Splash loading screen. How can I fix this problem? 
Thanks.

**Comment by rtivital** [comment]
You did not configure ssr properly. Follow Next.js guide to setup ssr or use template, it does not have any issues – https://github.com/mantinedev/mantine-next-template

**Reply by EliseiNicolae** [reply]
Hi @rtivital, currently that page is unavailable :)

**Comment by ehsanonline** [comment, marked as answer]
Thanks. It seems the problem was with `emotionOptions`. I'm using RTL support and [as doc said](https://mantine.dev/theming/rtl/#rtl-with-nextjs), I've added `emotionOptions={{ key: 'rtl', stylisPlugins: [rtlPlugin] }}` to the provider, which makes the FOUC problem. but when I changed the `key` from `rtl` to `mantine` the problem solved.

**Reply by ehsanonline** [reply]
this is not working with 4.1 update, looking for a new solution...

**Reply by ehsanonline** [reply]
problem solved by updating `@mantine/next` and `@mantine/styles` to 4.1 manually.

**Comment by ehsanonline** [comment]
Having this issue again, even without RTL. I created new next app and tried all the tutorials in the docs, I think it's a bug now. 
reproduce : build the next project and start it. open the page in chrome, open dev tools, set throttling on slow 3G at Network tab, keep dev tools open, hold refresh button then click Hard Reload.

![mantine-fouc](https://user-images.githubusercontent.com/4243503/168467887-1c18c6fa-28b6-471a-b685-65dcaec5b7b3.jpg)

**Reply by rtivital** [reply]
Custom cache key is not currently supported for ssr (emotion options), I'll provide a way to fix that in 5.0 since it will be a breaking change

**Comment by EliseiNicolae** [comment]
//EDIT

For those who have the same issue in Next.JS, check if you configured correctly the `mantine` Provider.
Doc: https://mantine.dev/guides/next/

**Reply by bismark66** [reply]
I am encountering this same problem with nextjs, i am talking about FOUC in production and i have done a lot and yet the problem is still there. Can someone help me out. I have used the provider configuration and yet the problem is there

**Reply by ttpss930141011** [reply]
Hi @bismark66 , I fixed the FOUC problem by implementing `RootStyleRegistry` mentioned by [here](https://github.com/emotion-js/emotion/issues/2928#issuecomment-1293012737)
Other reference:
- [Next.js Dcoument](https://nextjs.org/docs/app/building-your-application/styling/css-in-js)
- https://github.com/emotion-js/emotion/issues/2928
- https://github.com/mantinedev/mantine/issues/2815

**Reply by emwalker** [reply]
What resolved the FOUC (flash of unstyled content) for me was to add a missing `<ColorSchemeScript />` [in the document `<head>` section](https://github.com/mantinedev/next-pages-template/blob/7e9b96e5cba43ba38b44093eeacb1148088fcf25/pages/_document.tsx#L8C9-L8C30).

**Comment by smyja** [comment]
> I am encountering this same problem with nextjs, i am talking about FOUC in production and i have done a lot and yet the problem is still there. Can someone help me out. I have used the provider configuration and yet the problem is there
I am also having the same issue, but I think the problem is from next js.
